title: "Service should recover from container failure"
description: |
  A chaos experiment to verify that a service can successfully recover
  from container failure within a certain time frame.

contributions:
  security: "low"
  reliability: "high"
  scalability: "medium"

configuration:
  cluster_name: "live-comments-full-cluster"
  service_name: "web-service"
  acceptable_downtime: 120
  application_url:
    type: probe
    provider:
      type: python
      module: devlearnops.aws.ssm.probes
      func: get_parameter
      arguments:
        name: "/app/comments/live/web_url"

steady-state-hypothesis:
  title: "service-is-stable"
  probes:
    - name: "service-responsive"
      type: probe
      tolerance: 200
      provider:
        type: http
        url: "${application_url}/users"
        method: GET
        timeout: 2
    - name: "desired-equals-running"
      type: probe
      tolerance: true
      provider:
        type: python
        module: chaosaws.ecs.probes
        func: are_all_desired_tasks_running
        arguments:
          cluster: ${cluster_name}
          service: ${service_name}

method:
  - name: "fail-half-containers"
    type: action
    provider:
      type: python
      module: chaosaws.ecs.actions
      func: stop_random_tasks
      arguments:
        cluster: ${cluster_name}
        service: ${service_name}
        task_percent: 50
        reason: "Chaos testing"
    pauses:
      after: ${acceptable_downtime}

